### PR TITLE
Migrating unittest to pytest 6

### PR DESCRIPTION
**Pull request by prady0t**
I've tried using fixture and parameterized in this PR. Do let me know places I've missed (inside this PR).

#4156

### CODE REVIEW

File: tests/unit/test_geometry/test_battery_geometry.py
Comment (agriyakhetarpal) [inline tests/unit/test_geometry/test_battery_geometry.py:27]: We initialise the `geometry` variable below, do we need it in the arguments of the function?

Comment (agriyakhetarpal) [reply tests/unit/test_geometry/test_battery_geometry.py:27]: Also, this test can be refactored since we are testing the same thing repeatedly. Please add this to the tracking issue.

File: tests/unit/test_meshes/test_scikit_fem_submesh.py
Comment (agriyakhetarpal) [inline tests/unit/test_meshes/test_scikit_fem_submesh.py:442]: This test can be refactored, too.

File: tests/unit/test_parameters/test_parameter_sets/test_OKane2022.py
Comment (agriyakhetarpal) [inline tests/unit/test_parameters/test_parameter_sets/test_OKane2022.py:7]: Looks great, definitely add more like these! 👍

File: tests/unit/test_models/test_full_battery_models/test_lithium_ion/test_electrode_soh.py
Comment (agriyakhetarpal) [inline tests/unit/test_models/test_full_battery_models/test_lithium_ion/test_electrode_soh.py:18]: Could you move this fixture down to just above `TestElectrodeSOHMSMR`, so that we know it is relevant only for those tests?

Comment (agriyakhetarpal) [reply tests/unit/test_models/test_full_battery_models/test_lithium_ion/test_electrode_soh.py:18]: Either that, or (maybe better) you can add in a comment here the names of the classes that this applies to.

File: tests/unit/test_models/test_base_model.py
Comment (agriyakhetarpal) [inline tests/unit/test_models/test_base_model.py:195]: Wherever you find repetition like this, it's probably better to parameterize the function and gain benefits, both in terms of readability and slightly faster execution by parallelization.

File: tests/unit/test_models/test_base_model.py
Comment (agriyakhetarpal) [inline tests/unit/test_models/test_base_model.py:288]: This looks like it can be parameterized as well.

File: tests/unit/test_models/test_base_model.py
Comment (agriyakhetarpal) [inline tests/unit/test_models/test_base_model.py:387]: This can be parameterized as well. I would recommend opening a subsequent PR that implements nothing but parameterization (and similarly for fixtures, etc.) and changing this test there.

File: tests/unit/test_expression_tree/test_variable.py
Comment (agriyakhetarpal) [inline tests/unit/test_expression_tree/test_variable.py:20]: Could you take a look at this failure? – I noticed from the annotation, which was quite helpful. It looks like a `pytestify` bug, where, instead of converting `assertRaises` to a

```python
with pytest.raises(TypeError):
    pybamm.Variable("a", domain="test")
```

it was converted to a regular test where `match` expects a string, but it received a `pybamm.Variable` instance, which won't work.

File: tests/unit/test_expression_tree/test_variable.py
Comment (agriyakhetarpal) [inline tests/unit/test_expression_tree/test_variable.py:79]: Same error as above here

File: tests/unit/test_expression_tree/test_variable.py
Comment (prady0t) [reply tests/unit/test_expression_tree/test_variable.py:20]: Also not sure if `TypeError` should be raised here anyways. See : https://github.com/pybamm-team/PyBaMM/blob/396741eec6076a83c3104aa9b73d2f7957201ab7/src/pybamm/type_definitions.py#L17
i.e. domain can be of type `string` or a list of it as well.

Comment (agriyakhetarpal) [reply tests/unit/test_expression_tree/test_variable.py:20]:
```python
self.assertRaises(TypeError, pybamm.Variable("a", domain="test"))
```

works fine, as you mentioned, but

even the `unittest` equivalent of the `pytest.raises` context manager, i.e.,

```python
with self.assertRaises(TypeError):
    pybamm.Variable("a", domain="test")
```

fails with the same error not being raised.

I don't think this test was written correctly, since we should always use a context manager to catch errors. Either that, or we never enforced the type in the first place. For example, here is a short snippet of what happens if we enforce the type and raise an error:

```python
from typing import TypeAlias, Union
import unittest
import pytest


DomainType: TypeAlias = Union[list[str], str, None]


def foo(name: str, domain: DomainType = None) -> tuple[str, DomainType]:
    if isinstance(domain, int):
        raise TypeError
    return (name, domain)

# both should work
foo(name="a", domain="test")
foo(name="a", domain=["test"])


class TestFoo(unittest.TestCase):
    def test_foo(self): # raises an error
        self.assertRaises(TypeError, foo(name="a", domain=1))
    def test_foo2(self): # passes
        with self.assertRaises(TypeError):
            foo(name="a", domain=1)
    def test_foo3(self): # passes
        with pytest.raises(TypeError):
            foo(name="a", domain=1)

```

Here, `self.assertRaises(TypeError, foo(name="a", domain=1))` raises an error instead of failing because `foo` already raises a `TypeError` at runtime, which can't be caught out of context. We could enforce this, but it would contradict the docstring of `pybamm.Variable()`, which says that the argument is "domain : iterable of str", and changing the behaviour of the function means this is a breaking change – not recommended and outside the scope of migrating our testing infrastructure. I think we should be okay with removing this test case, since the behaviour is clearly correct – it shouldn't raise an error since we don't validate the type in the code. Do you have any thoughts, @Saransh-cpp?

Comment (prady0t) [reply tests/unit/test_expression_tree/test_variable.py:20]: Totally agreed. We should remove this and the following test. Waiting on Saransh reply.

Comment (Saransh-cpp) [reply tests/unit/test_expression_tree/test_variable.py:20]: Removing the test case sounds good to me too

File: tests/unit/test_geometry/test_battery_geometry.py
Comment (prady0t) [reply tests/unit/test_geometry/test_battery_geometry.py:27]: Now sure how to refactor. Could you help me more?

Comment (agriyakhetarpal) [reply tests/unit/test_geometry/test_battery_geometry.py]: Usually, in such situations, it's better to make the test shorter and distribute it into multiple tests. For example, here, you could rewrite `test_geometry` into an array of smaller tests (haven't added the `self` keyword here for brevity):

```python
def test_battery_geometry_current_collector_tabs():
    geometry = pybamm.battery_geometry(
        options={
            "particle size": "distribution",
            "dimensionality": 1,
        },
    )
    assert "tabs" in geometry["current collector"]
```

and

```python
def test_battery_geometry_cylindrical_dimensionality():
    geo = pybamm.geometric_parameters
    geometry = pybamm.battery_geometry(
        form_factor="cylindrical", options={"dimensionality": 1}
    )
    assert geometry["current collector"]["r_macro"]["min"] == geo.r_inner
    assert geometry["current collector"]["r_macro"]["max"] == 1
```

However, readability is important. So, if refactoring introduces too many conditions or makes the code hard to read, it is best to not bother and keep the test as is.

Another aspect is to remove the loops by parametrizing. For example, the loop `for cc_dimension in [0, 1, 2]:` can be refactored into

```python
@pytest.mark.parametrize("cc_dimension", [0, 1, 2])
def test_battery_geometry_basics(cc_dimension):
```

and similarly

```python
@pytest.mark.parametrize("electrode, particle_type", [
    ("negative", "primary"),
    ("negative", "secondary"),
    ("positive", "primary"),
    ("positive", "secondary"),
])
def test_battery_geometry_particle_phases(electrode, particle_type):
    options = {"particle phases": "2"}
    geometry = pybamm.battery_geometry(options=options)
    geo = pybamm.GeometricParameters(options=options)
# followed by the rest of the test
```

and so on. I hope this can get you going!

Comment (agriyakhetarpal) [reply tests/unit/test_geometry/test_battery_geometry.py]: I'm happy to write more examples if needed, it's easier than you would think :)

Comment (arjxn-py) [reply tests/unit/test_geometry/test_battery_geometry.py]: These examples looks enough to me. Thanks :)
I guess we don't need case or if/else then as we using `parametrize` marker?

Comment (agriyakhetarpal) [reply tests/unit/test_geometry/test_battery_geometry.py]: In some cases, we might still need them.

```python

@pytest.mark.parametrize("foo, bar", [0, 1, 2, 3, 4])
def test_something(foo, bar):
    # Some logic for initialising "foo" and "bar" here
    # and setting attributes, etc.
    # [...]

    # Conditional logic
    match foo.value:
        case 1:
            # do something
            assert foo in bar
        case 3.value:
            # do something else
            assert bar in foo

    # not (1) and not (3) – i.e., for (0) or (2) or (4)
    # Some other logic that applies to the rest of the cases
    assert foo.value + bar.value > 30
```

but it depends on the test. It _might_ be better to handle such cases without parameterizing them

**Comment by prady0t** [comment]
Not sure why the error is. I was getting another regex error locally can you run CI jobs again?

File: tests/unit/test_parameters/test_parameter_sets/test_OKane2022.py
Comment (arjxn-py) [inline tests/unit/test_parameters/test_parameter_sets/test_OKane2022.py:8]:
```suggestion
        {
```

**Comment by agriyakhetarpal** [comment]
Since this PR seems to have taken a bit of time (for all the right reasons), let's nudge this towards a merge asap and work on the parameterization in further, dedicated PRs in your coming weeks – once the `pytest` migration gets finished.

File: tests/unit/test_parameters/test_parameter_sets/test_OKane2022.py
Comment (Saransh-cpp) [inline tests/unit/test_parameters/test_parameter_sets/test_OKane2022.py:6]: The `(` is not close below

**Comment by prady0t** [comment]
I reverted the changes in `test_OKane2022.py` I still need to figure out the regex error. In the meantime, I'm also working on the last migration. I'll try to open PR by today or tomorrow morning.

**Comment by prady0t** [comment]
I've removed pytest conversion from `test_base_battery_model.py`. This should work

File: test_spm_2024_09_13-PM03_52.json
Comment (agriyakhetarpal) [inline test_spm_2024_09_13-PM03_52.json:1]: This generated test file has crept into the changes for some reason, it should be removed

**Review by agriyakhetarpal — CHANGES_REQUESTED** [review]
Looks nice, thanks, @prady0t! Just one more change, and we can merge this if the tests pass.

File: test_spm_2024_09_13-PM03_52.json
Comment (prady0t) [reply test_spm_2024_09_13-PM03_52.json:1]: Ah almost forgot about this

**Review by agriyakhetarpal — APPROVED** [review]

**Pull request merged**
